src/common/handler-error.ts

```diff
@@ -1,100 +1,201 @@
- * this request unless advised otherwise.
+ * ```ts
+ *     import { HandlerError } from "@nexus-rpc/sdk-typescript";
  *
- * RESOURCE_EXHAUSTED - Some resource has been exhausted, perhaps a per-user quota, or perhaps the entire file system
- * is out of space. Subsequent requests by the client are permissible.
+ *     // Throw a bad request error
+ *     throw new HandlerError("BAD_REQUEST", "Invalid input provided");
  *
- * INTERNAL - An internal error occured. Subsequent requests by the client are permissible.
+ *     // Throw a bad request error, with a cause
+ *     throw new HandlerError("BAD_REQUEST", "Invalid input provided", { cause });
  *
- * NOT_IMPLEMENTED - The server either does not recognize the request method, or it lacks the ability to fulfill the
- * request. Clients should not retry this request unless advised otherwise.
+ *     // Throw a retryable internal error
+ *     throw new HandlerError("INTERNAL", "Database unavailable", { retryableOverride: true });
+ * ```
  *
- * UNAVAILABLE - The service is currently unavailable. Subsequent requests by the client are permissible.
- *
- * UPSTREAM_TIMEOUT - Used by gateways to report that a request to an upstream server has timed out. Subsequent
- * requests by the client are permissible.
+ * @experimental
  */
-export type HandlerErrorType =
-  | "BAD_REQUEST"
-  | "UNAUTHENTICATED"
-  | "UNAUTHORIZED"
-  | "NOT_FOUND"
-  | "RESOURCE_EXHAUSTED"
-  | "INTERNAL"
-  | "NOT_IMPLEMENTED"
-  | "UNAVAILABLE"
-  | "UPSTREAM_TIMEOUT";
+export class HandlerError extends Error {
+  /**
+   * One of the predefined error types.
+   *
+   * @see {@link HandlerErrorType}
+   */
+  public readonly type: HandlerErrorType;
 
-/**
- * Options for constructing a {@link HandlerError} from a message, type, and a retryable flag.
- */
-export interface HandlerErrorMessageOptions {
-  /** Error message. */
-  message: string;
+  /**
+   * Whether this error should be considered retryable.
+   *
+   * By default, the retry behavior is determined from the error type.
+   * For example, by default, `INTERNAL` is retryable, but `UNAVAILABLE` is non-retryable.
+   *
+   * If specified, `retryableOverride` overrides the default retry behavior determined based on
+   * the error type. Use {@link retryable} to determine the effective retry behavior.
+   *
+   * @see {@link retryable}.
+   */
+  public readonly retryableOverride: boolean | undefined;
 
-  /** One of the predefined error types. */
-  type: HandlerErrorType;
+  /**
+   * Constructs a new {@link HandlerError}.
+   *
+   * @param type - The type of the error.
+   * @param message - The message of the error.
+   * @param options - Extra options for the error, including the cause and retryable override.
+   *
+   * @experimental
+   */
+  constructor(type: HandlerErrorType, message?: string | undefined, options?: HandlerErrorOptions) {
```

Comment:
Also seems redundant:

```suggestion
  constructor(type: HandlerErrorType, message?: string, options?: HandlerErrorOptions) {
```